substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/configure/LegacyResourceConfigurationParser.java

```diff
@@ -0,0 +1,118 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.configure;
+
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import org.graalvm.collections.EconomicMap;
+import org.graalvm.collections.MapCursor;
+import org.graalvm.nativeimage.impl.ConfigurationCondition;
+
+import com.oracle.svm.core.TypeResult;
+
+final class LegacyResourceConfigurationParser extends ResourceConfigurationParser {
+    LegacyResourceConfigurationParser(ConfigurationConditionResolver conditionResolver, ResourcesRegistry registry, boolean strictConfiguration) {
+        super(conditionResolver, registry, strictConfiguration);
+    }
+
+    @Override
+    public void parseAndRegister(Object json, URI origin) {
+        parseTopLevelObject(asMap(json, "first level of document must be an object"));
+    }
+
+    @Override
+    protected ConfigurationCondition parseCondition(EconomicMap<String, Object> data) {
+        return parseCondition(data, false);
+    }
+
+    private void parseTopLevelObject(EconomicMap<String, Object> obj) {
+        Object resourcesObject = null;
+        Object bundlesObject = null;
+        Object globsObject = null;
+        MapCursor<String, Object> cursor = obj.getEntries();
+        while (cursor.advance()) {
+            if (RESOURCES_KEY.equals(cursor.getKey())) {
+                resourcesObject = cursor.getValue();
+            } else if (BUNDLES_KEY.equals(cursor.getKey())) {
+                bundlesObject = cursor.getValue();
+            } else if (GLOBS_KEY.equals(cursor.getKey())) {
+                globsObject = cursor.getValue();
+            }
+        }
+
+        if (resourcesObject != null) {
+            parseResourcesObject(resourcesObject);
+        }
+        if (bundlesObject != null) {
+            parseBundlesObject(bundlesObject);
+        }
+        if (globsObject != null) {
+            parseGlobsObject(globsObject);
+        }
+    }
```

Comment:
If I get this right, you need to partially backport globs support for `reachability-metadata`, but do we really want to backport globs support to `resource-config` as well?

Reply:
Globs can be present in a `resource-config.json` file in the case where both globs and resource patterns are registered in the same `ResourceConfiguration` (for example when merging both combined and legacy metadata files). In that case JSON printing falls back to emitting a standalone legacy file instead of integrating resources in `reachability-metadata.json` (where we want to avoid resource patterns moving forwards), which means we need to be able to emit globs in this legacy file.

> If we do, the https://github.com/oracle/graal/commit/17fc7f0b0f7eda7d50ed349b337fae375e8aef1e implementation be preferred?

This change applies to JSON emission to ensure that converting legacy `resource-config.json` files to `reachability-metadata.json` succeeds most of the time instead of having to fallback to `resource-config.json` because it contains resource patterns. Since the Java 21 version of Graal doesn't emit `reachability-metadata.json` files, this optimization is not necessary.

Reply:
> Globs can be present in a `resource-config.json` file 

Not in JDK 21 (GraalVM 23.1), that feature was added in JDK 23 (GraalVM 24.1.0). So what I am worried about is backporting the globs support without actually intending to do so. 

As you say in https://github.com/graalvm/graalvm-community-jdk21u/pull/23#discussion_r1775323595:

> We do not want to backport new builder functionality to previous versions, but only allow previous versions to correctly parse newer reachability-metadata.json files.

Regarding:

> Since the Java 21 version of Graal doesn't emit reachability-metadata.json files, this optimization is not necessary.

I see, what I worry about though is making future backports harder due to deviating from the actual upstream code (which we already do by not fully backporting https://github.com/oracle/graal/pull/8250)

Reply:
> But metadata files created on GraalVM 24.1 can be passed to GraalVM 23.1 and are expected to be understood.

That would be in the `reachability-metadata.json` format though, not in the legacy `resource-config.json` format, right? 

If we want to be able to consume "legacy" config files **generated by newer GraalVM versions** then we need to also support `type` in `reflect-config.json` and `proxies` in `serialization-config.json`

> Since globs are meant as a replacement for resource patterns and as such are not purely a new capability of the builder, backporting glob parsing is necessary, otherwise resource handling through reachability-metadata.json would become impossible on 23.1.

I agree, my only concern is about bringing support for globs in `resource-config.json` **in addition to** `reachability-metadata.json`.

> That's a fair point, keeping parser code as close to master as possible should indeed be a goal. I will start the backporting process for this change and either include it in this PR or open a new one once it's done.

Thank you

Reply:
@loicottet can you please have a look in the above comments and provide some more details? It's still not clear to me at what level we need to support "globs support".